rdm/0007-access-control.md

```diff
@@ -0,0 +1,222 @@
+        return Document._owners
+    excludes:
+        pass
+```
+
+Since the record owner does not set any specific blacklist (*exclude*), the generator would result in an access granted for users with id 1, 2 or 3.
+
+In addition, generators resturn a set of *query filters*, which are used only when searching. The set of rules returned there are in the form of an Elasticsearch Query object. The `RecordOwner` generator would therefore look like:
+
+``` python
+RecordOwner:
+    needs:
+        return Document._owners
+    excludes:
+        pass
+    query_filter:
+        return Query(owner=user.id)
+```
+
+As a result, it can be seen that the *query filter* inverts the paradigm with respect to the *needs*/*excludes*. Instead of receiving a record and returning its owners, it takes the id of the user performing the action and returns a query object. This is due to the nature of the *list* (search) operation. The check must not be *Can user A access record R, for each one of the records in the system*. This would not be scalable. Therefore, the check must be *which documents can be accessed by the user A*.
+
+Invenio RDM comes with a set of predefined generators:
+- Any user: Any user can perform the action.
+- Any user if public: Any user can perform the action if the record is public.
+- Record owners: Owners of the record can access perform the action.
+- Curators: Curators can perform an action.
+- Super Curators: ??
+
+
+**Permission Policies**
+
+A *permission policy* defines the set of generators upon which a certain action can be performed. A record policy would be set as:
+
+``` python
+RecordPermissionPolicy:
+    can_create: []
+    can_read: []
+    can_update: []
+    can_delete: []
+    can_list: []
+```
+
+The policy would return a set of *needs* and *excludes* for each action. The *needs* and *excludes* of a policy is the union of all those of the specified generators.
+
+For the following user story:
+
+```
+As an admin I want to grant read access to the owners of a document, except if the owner is part of team A.
+```
+
+The following policy applies the `RecordOwner` and `TeamMembership` generators to the record read action. The `RecordOwner` generator, allows the owners of the document (defined right above it, with ids 1, 2 and 3) to access the document. On the other hand, `TeamMembershipt` blacklists users belonging to Team A of accesing the document.
+
+``` python
+Document:
+    _owners: [1, 2, 3]
+
+RecordPermissionPolicy:
+    can_read: [RecordOwner, TeamMembership]
+
+RecordOwner:
+    needs:
+        return Document._owners
+
+TeanMembership:
+    needs:
+        pass
+    excludes:
+        return [Team A]
+```
+
+Assuming the following two users:
+
+``` python
+UserOne:
+    id: 1
+    team: B
+
+UserTwo:
+    id: 2
+    team: A
+```
+
+*UserOne* would be able to read the document since it has id `1`. However, *UserTwo* would not. Eventhough, it has id `2`, it belongs to `Team A` which has been blacklisted.
+
```

Comment:
Having a concrete example like this was really helpful. I am tempted to add  **Example** section to `template.md`.

Reply:
I agree, however, this is only possible on more "implementation" rfcs. Maybe we can add it as optional?